src/main/java/core/basesyntax/strategy/PurchaseOperation.java

```
@@ -0,0 +1,12 @@
+package core.basesyntax.strategy;
+
+import core.basesyntax.common.FruitTransaction;
+import core.basesyntax.db.Storage;
+
+public class PurchaseOperation implements OperationHandler {
+    @Override
+    public void apply(FruitTransaction fruitTransaction) {
+        Storage.storage.merge(fruitTransaction.getFruit(),
+                -fruitTransaction.getQuantity(), Integer::sum);
```

Handlers must validate their inputs before applying changes. This `apply` method lacks checks that `fruitTransaction` is non-null, that `fruitTransaction.getFruit()` is non-null/non-empty (after trim), and that `fruitTransaction.getQuantity()` is non-negative. The project requirements state handlers should validate inputs (non-null transaction, fruit non-null/non-empty, and non-negative quantity). Add those validations and throw informative RuntimeExceptions for invalid input rather than allowing NPEs or silent incorrect behavior. (The problematic call site is lines 8–10 where values are directly used.)